pkg/redisx/pool.go

```diff
@@ -0,0 +1,165 @@
+package redisx
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/goharbor/harbor-scanner-clair/pkg/etc"
+
+	"github.com/FZambia/sentinel"
+	"github.com/gomodule/redigo/redis"
+	log "github.com/sirupsen/logrus"
+)
+
+// NewPool constructs a redis.Pool for the given RedisPool config params.
+//
+// The currently supported Redis URL schemas correspond to a Redis server
+// run in standalone and Sentinel modes:
+// - redis://user:password@standalone_host:port/db-number
+// - redis+sentinel://user:password@sentinel_host1:port1,sentinel_host2:port2/monitor-name/db-number
+func NewPool(config etc.RedisPool) (*redis.Pool, error) {
+	configURL, err := url.Parse(config.URL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid redis URL: %s", err)
+	}
+
+	switch configURL.Scheme {
+	case "redis":
+		return newInstancePool(config), nil
+	case "redis+sentinel":
+		return newSentinelPool(configURL, config)
+	default:
+		return nil, fmt.Errorf("invalid redis URL scheme: %s", configURL.Scheme)
+	}
+}
+
+// redis://user:password@standalone_host:port/db-number
+func newInstancePool(config etc.RedisPool) *redis.Pool {
+	log.Trace("Constructing connection pool for Redis instance")
+	return &redis.Pool{
+		Dial: func() (redis.Conn, error) {
+			log.WithField("url", config.URL).Trace("Connecting to Redis")
+			return redis.DialURL(config.URL)
+		},
+		TestOnBorrow: func(c redis.Conn, t time.Time) error {
+			if time.Since(t) < time.Minute {
+				return nil
+			}
+			log.Trace("Testing connection to Redis on borrow")
+			_, err := c.Do("PING")
+			return err
+		},
+		MaxIdle:     config.MaxIdle,
+		MaxActive:   config.MaxActive,
+		IdleTimeout: config.IdleTimeout,
+		Wait:        true,
+	}
+}
+
+// redis+sentinel://user:password@sentinel_host1:port1,sentinel_host2:port2/monitor-name/db-number
+func newSentinelPool(configURL *url.URL, config etc.RedisPool) (pool *redis.Pool, err error) {
+	log.Trace("Constructing connection pool for Redis Sentinel")
+	sentinelURL, err := ParseSentinelURL(configURL)
+	if err != nil {
+		return
+	}
+
+	var commonOpts []redis.DialOption
+	if config.ConnectionTimeout > 0 {
+		commonOpts = append(commonOpts, redis.DialConnectTimeout(config.ConnectionTimeout))
+	}
+	if config.ReadTimeout > 0 {
+		commonOpts = append(commonOpts, redis.DialReadTimeout(config.ReadTimeout))
+	}
+	if config.WriteTimeout > 0 {
+		commonOpts = append(commonOpts, redis.DialWriteTimeout(config.WriteTimeout))
+	}
+
+	sentinelOpts := commonOpts
+
+	sntnl := &sentinel.Sentinel{
+		Addrs:      sentinelURL.Addrs,
+		MasterName: sentinelURL.MonitorName,
+		Dial: func(addr string) (conn redis.Conn, err error) {
+			log.WithField("addr", addr).Trace("Connecting to Redis sentinel")
+			conn, err = redis.Dial("tcp", addr, sentinelOpts...)
+			if err != nil {
```

Comment:
I think one return is ok

Reply:
Good point @steven-zou I simplified the return statements here